File: android/dotrino-native/src/main/java/com/dotrino/sdk/AccountStore.kt
package com.dotrino.sdk

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import java.io.File
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * The accounts this phone approves for, on disk. Nothing secret is stored (vault key, proxy,
 * paper, name), but the list of vaults a person belongs to is theirs, so the file is
 * encrypted with an AES key from the Keystore that never leaves it.
 *
 * A file that exists but does not open is an ERROR, not an empty list: treating it as
 * «no accounts» would silently make the phone stop approving.
 */
class AccountStore(context: Context) {
    private val file = File(context.filesDir, "dotrino-accounts.bin")
    private val json = Json { ignoreUnknownKeys = true }
    private val ser = ListSerializer(Account.serializer())

    companion object {
        private const val ALIAS = "dotrino.accounts"
        private fun key(): SecretKey {
            val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
            (ks.getKey(ALIAS, null) as? SecretKey)?.let { return it }
            val spec = KeyGenParameterSpec.Builder(ALIAS, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .build()
            return KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore").apply { init(spec) }.generateKey()
        }
    }

    @Synchronized fun list(): List<Account> {
        if (!file.exists()) return emptyList()
        val raw = file.readBytes()
        require(raw.size > 12) { "accounts file is truncated" }
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.DECRYPT_MODE, key(), GCMParameterSpec(128, raw.copyOfRange(0, 12)))
        return json.decodeFromString(ser, String(c.doFinal(raw.copyOfRange(12, raw.size)), Charsets.UTF_8))
    }

    @Synchronized private fun write(accounts: List<Account>) {
        val c = Cipher.getInstance("AES/GCM/NoPadding")
        c.init(Cipher.ENCRYPT_MODE, key())
        val body = c.iv + c.doFinal(json.encodeToString(ser, accounts).toByteArray(Charsets.UTF_8))
        // Write to a sibling and rename: a half-written file must never replace a good one.
        val tmp = File(file.parentFile, file.name + ".tmp")
        tmp.writeBytes(body)
        if (!tmp.renameTo(file)) throw IllegalStateException("could not save the accounts file")
    }

    /** Adds or replaces the account with the same [Account.id]. */
    @Synchronized fun save(a: Account) = write(list().filter { it.id != a.id } + a)

    @Synchronized fun remove(id: String) {
        write(list().filter { it.id != id })
        KeystoreKeys.delete(id)
    }
}

File: ios/DotrinoNative/AccountStore.swift
import Foundation

/// An account this phone approves for natively: which vault, through which proxy, and the
/// paper (`cert`) the vault gave to this device's key. No secret in here — the keys live in
/// the Secure Enclave under [id].
public struct Account: Equatable, Sendable {
    public let id: String
    public let name: String
    public let profileId: String?
    public let vault: String
    public let proxy: String
    public var cert: JSON
    public let deviceId: String
    public let addedAt: Int64

    public init(id: String, name: String, profileId: String?, vault: String, proxy: String, cert: JSON, deviceId: String, addedAt: Int64 = nowMs()) {
        self.id = id; self.name = name; self.profileId = profileId; self.vault = vault
        self.proxy = proxy; self.cert = cert; self.deviceId = deviceId; self.addedAt = addedAt
    }

    var json: JSON {
        var o: [String: JSON] = ["id": .string(id), "name": .string(name), "vault": .string(vault), "proxy": .string(proxy),
                                 "cert": cert, "deviceId": .string(deviceId), "addedAt": .int(addedAt)]
        if let profileId { o["profileId"] = .string(profileId) }
        return .object(o)
    }

    init(json o: JSON) throws {
        guard let id = o["id"]?.string, let name = o["name"]?.string, let vault = o["vault"]?.string,
              let proxy = o["proxy"]?.string, let cert = o["cert"], cert.object != nil, let deviceId = o["deviceId"]?.string
        else { throw CryptoError("accounts file: an account is missing fields") }
        self.init(id: id, name: name, profileId: o["profileId"]?.string, vault: vault, proxy: proxy, cert: cert,
                  deviceId: deviceId, addedAt: o["addedAt"]?.int ?? 0)
    }
}

/// The accounts this phone approves for, on disk, in a [SealedFile]. Nothing secret is
/// stored (vault key, proxy, paper, name), but the list of vaults a person belongs to is theirs.
public final class AccountStore: @unchecked Sendable {
    public static let shared = AccountStore()
    private let sealed = SealedFile(name: "dotrino-accounts.bin", keyAlias: "dotrino.accounts")

    public init() {}

    public func list() throws -> [Account] {
        sealed.lock.lock(); defer { sealed.lock.unlock() }
        return try read()
    }

    private func read() throws -> [Account] {
        guard let plain = try sealed.read() else { return [] }
        guard let arr = try JSON.parse(plain).array else { throw CryptoError("accounts file is not a list") }
        return try arr.map(Account.init(json:))
    }

    private func write(_ accounts: [Account]) throws {
        try sealed.write(Data(JSON.array(accounts.map(\.json)).text.utf8))
    }

    /// Adds or replaces the account with the same [Account.id].
    public func save(_ a: Account) throws {
        sealed.lock.lock(); defer { sealed.lock.unlock() }
        try write(read().filter { $0.id != a.id } + [a])
    }

    public func remove(_ id: String) throws {
        sealed.lock.lock(); defer { sealed.lock.unlock() }
        try write(read().filter { $0.id != id })
        EnclaveKeys.delete(id)
    }
}
